### PR TITLE
Rewrote script (checked with www.shellcheck.net) to support more/diff…

### PR DESCRIPTION
**Pull request by cmanley**
…erent command-line options and unix_socket authentication.

### CODE REVIEW

File: shrink [hunk 1]
```diff
@@ -1,82 +1,269 @@
...
+
+
+# Check that required binaries are available.
+if [ -z "$(command -v mysql)" ]; then
+	die $? 'MySQL client (mysql) is not installed!'
+fi
+if [ -z "$(command -v mysqldump)" ]; then
+	die $? 'mysqldump not found!'
+fi
+GZIP="$(command -v pigz)"
```
Comment (uberhacker) [inline shrink]: Should this be `GZIP="$(command -v gzip)"`?

File: shrink [hunk 2]
```diff
@@ -1,82 +1,269 @@
...
+    warn "$@"
+    exit "$rc"
+}
+usage() {
+	msg="$*"
+	this=$(basename "$0")
+	if [ -n "$msg" ]; then
+		warn "$this: ERROR: $msg"
+	fi
+	warn "Usage: $this [-dk]"
```
Comment (uberhacker) [inline shrink]: `warn "Usage: $this [-cdhk]"`?

File: shrink [hunk 3]
```diff
@@ -1,82 +1,269 @@
...
+	warn '	Arguments:'
+	warn "		-c or --continue: Continue after failure. This restores all previously made *.sql.gz files too."
+	warn "		-d or --dry-run: Do not drop databases, remove files from $MYSQL_DATA_DIR, or restore databases."
```
Comment (uberhacker) [inline shrink]: Was the `-k` option intentionally removed?

File: shrink [hunk 4]
```diff
@@ -1,82 +1,269 @@
...
+	if [ -f "$FILE" ]; then
+		printf 'Restore %s to database %s... ' "$FILE" "$DATABASE"
+		if [ -z "$DRYRUN" ]; then
+			if ! "$GZIP" -dc "$FILE" | mysql -u"$MYSQL_USER" -p"$MYSQL_PASS"; then
+				die $? 'ERROR'
+			fi
+			echo 'OK'
+			rm -f "$FILE"
+		else
+			echo '(Skipped in DRY RUN)'
```
Comment (uberhacker) [inline shrink]: Would it be good to explain what exactly was skipped in the dry run?

File: shrink [hunk 5]
```diff
@@ -1,82 +1,269 @@
...
+		fi
+	fi
+done
+
+
+echo "Size of $IBDATA1_FILE before shrinking: $SIZE_BEFORE"
+SIZE_AFTER=$(du -h "$IBDATA1_FILE" | cut -f1)
+echo "Size of $IBDATA1_FILE after shrinking: $SIZE_AFTER"
```
Comment (uberhacker) [inline shrink]: No calculation of space recovered?

File: shrink [hunk 6]
```diff
@@ -1,82 +1,269 @@
...
+# This line is the only bash dependency:
+if [ "$(basename "$SHELL")" = 'bash' ]; then
+	set -o pipefail
+fi
+
+
+# Make sure mysqld is running
+if [ -z "$(pidof mysqld)" ]; then
+	echo "Starting mysql service."
+	service mysql start || die 1 "Failed to start mysql: $?"
```
Comment (uberhacker) [inline shrink]: This could be `service myslqd start` on some servers.  In fact, it might even need `systemctl start mysqld.service`.

**Comment by uberhacker** [comment]
@cmanley: Well...I probably should tag releases now. :)  Thank you for the rewrite.  I have added a few comments.

File: shrink (hunk 2 above)
Comment (cmanley) [reply shrink]: This has been fixed/updated in my latest commit:
https://github.com/cmanley/shrink-ibdata1/commit/4804e224cf24a358e9945da76b9585b811e96c24

**Comment by uberhacker** [comment]
Nice work @cmanley!

**Pull request merged**